src/net_processing.cpp

```diff
@@ -1142,60 +1143,71 @@ void static ProcessGetBlockData(CNode* pfrom, const Consensus::Params& consensus
         std::shared_ptr<const CBlock> pblock;
         if (a_recent_block && a_recent_block->GetHash() == pindex->GetBlockHash()) {
             pblock = a_recent_block;
+        } else if (inv.type == MSG_WITNESS_BLOCK) {
```

Comment:
Yes I'm not convinced this logic is correct. It seems to work, though, even for the initial blocks.

Reply:
I guess the only thing required would be to squeeze in the 0x0001 ("marker"+"flag"). This is identical to the serialization when calculating the hash (see https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#transaction-id)

Reply:
@laanwj It's always correct to give the raw blocks we store to peers that ask for witnesses (even if the block does not have a witness).

Converting extended format to basic format is a lot more complicated. You could have a special CTransaction which skips the witness fields instead of reading/deserializing them, but I don't see how to do it without going through some form of serialization code.

Reply:
Leaving out the less common edge case (non-witness peers) seems fine for now.

Reply:
Unsure if we care, but we could also check whether SegWit was active for the requested block, and if not, we can serve without deserialization even when witnesses are not requested.

Reply:
@sipa Yes, that would be something that could be done here.